src/tabmat/formula.py

```diff
@@ -0,0 +1,313 @@
+            return other.__mul__(self)
+        elif isinstance(other, (int, float)):
+            return self.multiply(numpy.array(other))
+        else:
+            raise TypeError(f"Cannot multiply {type(self)} and {type(other)}")
+        # Multiplication with categorical is handled by the categorical
+
+    def __rmul__(self, other):
+        return self.__mul__(other)
+
+    def to_non_interactable(self):
+        return SparseMatrix(self)
+
+    def get_names(self, col):
+        return [col]
+
+
+class InteractableCategoricalMatrix(CategoricalMatrix):
+    def __init__(self, *args, **kwargs):
+        multipliers = kwargs.pop("multipliers", None)
+        super().__init__(*args, **kwargs)
+        if multipliers is None:
+            self.multipliers = numpy.ones_like(self.cat, dtype=numpy.float_)
+        else:
+            self.multipliers = multipliers
+
+    def __mul__(self, other):
+        if isinstance(other, (InteractableDenseMatrix, float, int)):
+            result = copy.copy(self)
+            result.multipliers = result.multipliers * numpy.array(other)
+            return result
+        elif isinstance(other, InteractableSparseMatrix):
+            result = copy.copy(self)
+            result.multipliers = result.multipliers * other.todense()
+            return result
+        elif isinstance(other, InteractableCategoricalMatrix):
+            return self._interact_categorical(other)
+        else:
+            raise TypeError(
+                f"Can't multiply InteractableCategoricalMatrix with {type(other)}"
+            )
+
+    def __rmul__(self, other):
+        if isinstance(other, InteractableCategoricalMatrix):
+            other._interact_categorical(self)  # order matters
+        else:
+            return self.__mul__(other)
+
+    def to_non_interactable(self):
+        if numpy.all(self.multipliers == 1):
+            return CategoricalMatrix(
+                self.cat,
+                drop_first=self.drop_first,
+                dtype=self.dtype,
+            )
+        else:
+            return SparseMatrix(
+                self.tocsr().multiply(self.multipliers[:, numpy.newaxis])
+            )
+
+    def _interact_categorical(self, other):
+        cardinality_self = len(self.cat.categories)
+
+        new_codes = other.cat.codes * cardinality_self + self.cat.codes
+
+        if self.drop_first:
+            new_codes[new_codes % cardinality_self == 0] = 0
+            new_codes -= new_codes // cardinality_self
+            self_slice = slice(1, None)
+        else:
+            self_slice = slice(None)
+
+        if other.drop_first:
+            new_codes -= (cardinality_self - 1)
+            new_codes[new_codes < 0] = 0
+            other_slice = slice(1, None)
+        else:
+            other_slice = slice(None)
+
+        new_categories = [
+            f"{self_cat}:{other_cat}"
+            for other_cat, self_cat in itertools.product(
```

Comment:
Further idea for a check: compare output to model matrix from a statsmodel (or other GLM package) tutorial with a complicated model matrix (e.g., this [one](https://www.statsmodels.org/dev/examples/notebooks/generated/glm_formula.html)), trying different options like sparse or dense, drop_first or not, full structural rank or not. Doesn't need to be a test because we probably don't want the dependency.

Reply:
`tabmat` does not handle two-sided formulas (should it? I can be convinced :)), so I think this test would fit better in `glum` when we implement the formula interface there. We already have `statsmodels` as a dev dependency there so it can even be an automatic unit test.

Reply:
We should focus on the right hand side of the formula. 

I agree with the dependency issue but, on the other hand, formulas are a tabmat functionality, so tests for formula outputs should be in this package. We could test against patsy here and against statsmodels in glum. Or we test only against formulaic and rely on formulaic's tests against patsy to close the loop (I haven't checked that there are any, but I hope that they exist).

Irrespective of whether this becomes a test (and in which library( in the end, it is useful to try these formulas on some data examples. These examples could also become tutorials or examples later on. Here is one based on the link above. It would be great if we could also find one for categorical data and interactions with them (we could use area codes from the insurance example, but very happy to consider other sugggestions).
```
# %%
import tabmat as tm
import statsmodels.api as sm
from patsy import dmatrices

# %%
star98 = sm.datasets.star98.load_pandas().data
star98.dtypes
# %%
formula = "LOWINC + PERASIAN + PERBLACK + PERHISP + PCTCHRT + PCTYRRND + PERMINTE*AVYRSEXP*AVSALK + PERSPENK*PTRATIO*PCTAF"
formula_2s = "NABOVE ~ " + formula
# %%
y, X = dmatrices(formula_2s, data=star98, return_type='dataframe')

# %%
X_tm = tm.from_formula(formula, star98, ensure_full_rank=True).A
assert X.shape == X_tm.shape
# %% patsy seems to shuffle columns and interactions
# Does formulaic have a different convention here or is this an idiosyncracy of the tabmat 
# implementation?
(X - X_tm).head()
# %% model trained with either design matrix should be equivalent
mod = sm.OLS(y, X)
res = mod.fit() 

mod_tm = sm.OLS(y, X_tm)
res_tm = mod_tm.fit() 

assert res.aic == res_tm.aic

# %%
```

Reply:
The insurance data would be great for this! It shows off features like converting a categoricals on the fly, mix of numerical and categorical columns, etc.

Reply:
> ```python
> # Does formulaic have a different convention here or is this an idiosyncracy of the tabmat 
> # implementation?
> ```

Yes, that's [formulaic using a different convention](https://matthewwardrop.github.io/formulaic/migration/). `tabmat.from_formula` output should not differ from `formulaic.model_matrix` output in anything other than the return type. There is a test for that [here](https://github.com/Quantco/tabmat/blob/ef84e7dc2e65295b467f8dee0ea84e20ef02b1d2/tests/test_formula.py#L389) and [here](https://github.com/Quantco/tabmat/blob/formula/tests/test_formula.py#L283)